scrapy/link.py

```diff
@@ -13,11 +13,10 @@ class Link(object):
     __slots__ = ['url', 'text', 'fragment', 'nofollow']
 
     def __init__(self, url, text='', fragment='', nofollow=False):
-        if isinstance(url, six.text_type):
+        if not isinstance(url, str):
             import warnings
-            warnings.warn("Do not instantiate Link objects with unicode urls. "
-                "Assuming utf-8 encoding (which could be wrong)")
-            url = url.encode('utf-8')
+            warnings.warn("Link urls must be str objects.")
+            url = to_native_str(url)
```

Comment:
I prefer to fail early if `url` is not str in python3 and for backwards compatibiilty only warn when it is `unicode` in python2.